# Your Title Here

**Name(s)**:

**Website Link**: (your website link)

In [67]:
import pandas as pd
import numpy as np
from pathlib import Path

import plotly.express as px
pd.options.plotting.backend = 'plotly'

# from dsc80_utils import * # Feel free to uncomment and use this.

## Step 1: Introduction

In [68]:
data = pd.read_csv('2022.csv')
data

/Users/camellee/miniforge3/envs/dsc80/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3508: DtypeWarning:

Columns (2) have mixed types.Specify dtype option on import or set low_memory=False.



gameid datacompleteness  \
0       ESPORTSTMNT01_2690210         complete   
1       ESPORTSTMNT01_2690210         complete   
2       ESPORTSTMNT01_2690210         complete   
3       ESPORTSTMNT01_2690210         complete   
4       ESPORTSTMNT01_2690210         complete   
...                       ...              ...   
148987       9687-9687_game_5          partial   
148988       9687-9687_game_5          partial   
148989       9687-9687_game_5          partial   
148990       9687-9687_game_5          partial   
148991       9687-9687_game_5          partial   

                                                url league  year   split  \
0                                               NaN   LCKC  2022  Spring   
1                                               NaN   LCKC  2022  Spring   
2                                               NaN   LCKC  2022  Spring   
3                                               NaN   LCKC  2022  Spring   
4                                               NaN   LCKC  2022  Spring   
...                                             ...    ...   ...     ...   
148987  https://lpl.qq.com/es/stats.shtml?bmid=9687   DCup  2022     NaN   
148988  https://lpl.qq.com/es/stats.shtml?bmid=9687   DCup  2022     NaN   
148989  https://lpl.qq.com/es/stats.shtml?bmid=9687   DCup  2022     NaN   
148990  https://lpl.qq.com/es/stats.shtml?bmid=9687   DCup  2022     NaN   
148991  https://lpl.qq.com/es/stats.shtml?bmid=9687   DCup  2022     NaN   

        playoffs                 date  game  patch  ...  opp_csat15  \
0              0  2022-01-10 07:44:08     1  12.01  ...       121.0   
1              0  2022-01-10 07:44:08     1  12.01  ...       100.0   
2              0  2022-01-10 07:44:08     1  12.01  ...       119.0   
3              0  2022-01-10 07:44:08     1  12.01  ...       149.0   
4              0  2022-01-10 07:44:08     1  12.01  ...        21.0   
...          ...                  ...   ...    ...  ...         ...   
148987         0  2022-12-27 12:43:43     5  12.23  ...         NaN   
148988         0  2022-12-27 12:43:43     5  12.23  ...         NaN   
148989         0  2022-12-27 12:43:43     5  12.23  ...         NaN   
148990         0  2022-12-27 12:43:43     5  12.23  ...         NaN   
148991         0  2022-12-27 12:43:43     5  12.23  ...         NaN   

       golddiffat15 xpdiffat15 csdiffat15 killsat15 assistsat15 deathsat15  \
0             391.0      345.0       14.0       0.0         1.0        0.0   
1             541.0     -275.0      -11.0       2.0         3.0        2.0   
2            -475.0      153.0        1.0       0.0         3.0        0.0   
3            -793.0    -1343.0      -34.0       2.0         1.0        2.0   
4             443.0     -497.0        7.0       1.0         2.0        2.0   
...             ...        ...        ...       ...         ...        ...   
148987          NaN        NaN        NaN       NaN         NaN        NaN   
148988          NaN        NaN        NaN       NaN         NaN        NaN   
148989          NaN        NaN        NaN       NaN         NaN        NaN   
148990          NaN        NaN        NaN       NaN         NaN        NaN   
148991          NaN        NaN        NaN       NaN         NaN        NaN   

       opp_killsat15 opp_assistsat15 opp_deathsat15  
0                0.0             1.0            0.0  
1                0.0             5.0            1.0  
2                3.0             3.0            2.0  
3                3.0             3.0            0.0  
4                0.0             6.0            2.0  
...              ...             ...            ...  
148987           NaN             NaN            NaN  
148988           NaN             NaN            NaN  
148989           NaN             NaN            NaN  
148990           NaN             NaN            NaN  
148991           NaN             NaN            NaN  

[148992 rows x 130 columns]

## Step 2: Data Cleaning and Exploratory Data Analysis

In [69]:
data['league'].unique()

array(['LCKC', 'LPL', 'NLC', 'LVP SL', 'PGC', 'UL', 'PRM', 'LCK', 'LFL',
       'LEC', 'LCS', 'LFL2', 'GLL', 'HM', 'ESLOL', 'EBL', 'LPLOL', 'PGN',
       'LCSA', 'DDH', 'TAL', 'TCL', 'CBLOL', 'LCO', 'LHE', 'GL', 'EL',
       'CBLOLA', 'LMF', 'VL', 'SL (LATAM)', 'LLA', 'HC', 'LDL', 'LJL',
       'PCS', 'VCS', 'UPL', 'LCL', 'NEXO', 'EUM', 'LAS', 'MSI', 'LJLA',
       'CT', 'WLDs', 'CDF', 'IC', 'DCup'], dtype=object)

Change the dtype of the column `patch`

In [70]:
data['patch'].unique()

array([12.01, 12.02, 12.03, 12.04, 12.05, 12.06, 12.07, 12.08, 12.09,
       12.1 , 12.11, 12.12, 12.13, 12.14, 12.15,   nan, 12.16, 12.17,
       12.18, 12.19, 12.2 , 12.21, 12.23])

In [71]:
data['patch'] = data['patch'].astype(str).str.replace(r'^12.1$', '12.10', regex=True).str.replace(r'^12.2$', '12.20', regex=True)

### Keep only Tier 1

In [72]:
tier1 = data[data['league'].isin(['LPL', 'LCK', 'LCS', 'LEC', 'PCS', 'VCS', 'CBLOL', 'LLA', 'LMS', 'LCO', 'LJL'])]
tier1.head(15)

gameid datacompleteness  \
24  8401-8401_game_1          partial   
25  8401-8401_game_1          partial   
26  8401-8401_game_1          partial   
27  8401-8401_game_1          partial   
28  8401-8401_game_1          partial   
29  8401-8401_game_1          partial   
30  8401-8401_game_1          partial   
31  8401-8401_game_1          partial   
32  8401-8401_game_1          partial   
33  8401-8401_game_1          partial   
34  8401-8401_game_1          partial   
35  8401-8401_game_1          partial   
48  8401-8401_game_2          partial   
49  8401-8401_game_2          partial   
50  8401-8401_game_2          partial   

                                            url league  year   split  \
24  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
25  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
26  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
27  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
28  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
29  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
30  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
31  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
32  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
33  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
34  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
35  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
48  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
49  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
50  https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   

    playoffs                 date  game  patch  ...  opp_csat15 golddiffat15  \
24         0  2022-01-10 09:24:26     1  12.01  ...         NaN          NaN   
25         0  2022-01-10 09:24:26     1  12.01  ...         NaN          NaN   
26         0  2022-01-10 09:24:26     1  12.01  ...         NaN          NaN   
27         0  2022-01-10 09:24:26     1  12.01  ...         NaN          NaN   
28         0  2022-01-10 09:24:26     1  12.01  ...         NaN          NaN   
29         0  2022-01-10 09:24:26     1  12.01  ...         NaN          NaN   
30         0  2022-01-10 09:24:26     1  12.01  ...         NaN          NaN   
31         0  2022-01-10 09:24:26     1  12.01  ...         NaN          NaN   
32         0  2022-01-10 09:24:26     1  12.01  ...         NaN          NaN   
33         0  2022-01-10 09:24:26     1  12.01  ...         NaN          NaN   
34         0  2022-01-10 09:24:26     1  12.01  ...         NaN          NaN   
35         0  2022-01-10 09:24:26     1  12.01  ...         NaN          NaN   
48         0  2022-01-10 10:09:22     2  12.01  ...         NaN          NaN   
49         0  2022-01-10 10:09:22     2  12.01  ...         NaN          NaN   
50         0  2022-01-10 10:09:22     2  12.01  ...         NaN          NaN   

   xpdiffat15 csdiffat15 killsat15 assistsat15 deathsat15 opp_killsat15  \
24        NaN        NaN       NaN         NaN        NaN           NaN   
25        NaN        NaN       NaN         NaN        NaN           NaN   
26        NaN        NaN       NaN         NaN        NaN           NaN   
27        NaN        NaN       NaN         NaN        NaN           NaN   
28        NaN        NaN       NaN         NaN        NaN           NaN   
29        NaN        NaN       NaN         NaN        NaN           NaN   
30        NaN        NaN       NaN         NaN        NaN           NaN   
31        NaN        NaN       NaN         NaN        NaN           NaN   
32        NaN        NaN       NaN         NaN        NaN           NaN   
33        NaN        NaN       NaN         NaN        NaN           NaN   
34        NaN        NaN       NaN         NaN        NaN           NaN   
35        NaN        Na

#### Player Data

In [73]:
tier1_player = tier1.iloc[[i for i in range(len(tier1)) if i % 12 in range(10)]]
tier1_player

gameid datacompleteness  \
24           8401-8401_game_1          partial   
25           8401-8401_game_1          partial   
26           8401-8401_game_1          partial   
27           8401-8401_game_1          partial   
28           8401-8401_game_1          partial   
...                       ...              ...   
133409  ESPORTSTMNT03_2961733         complete   
133410  ESPORTSTMNT03_2961733         complete   
133411  ESPORTSTMNT03_2961733         complete   
133412  ESPORTSTMNT03_2961733         complete   
133413  ESPORTSTMNT03_2961733         complete   

                                                url league  year   split  \
24      https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
25      https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
26      https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
27      https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
28      https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
...                                             ...    ...   ...     ...   
133409                                          NaN    LLA  2023     NaN   
133410                                          NaN    LLA  2023     NaN   
133411                                          NaN    LLA  2023     NaN   
133412                                          NaN    LLA  2023     NaN   
133413                                          NaN    LLA  2023     NaN   

        playoffs                 date  game  patch  ...  opp_csat15  \
24             0  2022-01-10 09:24:26     1  12.01  ...         NaN   
25             0  2022-01-10 09:24:26     1  12.01  ...         NaN   
26             0  2022-01-10 09:24:26     1  12.01  ...         NaN   
27             0  2022-01-10 09:24:26     1  12.01  ...         NaN   
28             0  2022-01-10 09:24:26     1  12.01  ...         NaN   
...          ...                  ...   ...    ...  ...         ...   
133409         0  2022-09-11 23:09:07     4  12.16  ...       149.0   
133410         0  2022-09-11 23:09:07     4  12.16  ...        88.0   
133411         0  2022-09-11 23:09:07     4  12.16  ...       132.0   
133412         0  2022-09-11 23:09:07     4  12.16  ...       124.0   
133413         0  2022-09-11 23:09:07     4  12.16  ...        21.0   

       golddiffat15 xpdiffat15 csdiffat15 killsat15 assistsat15 deathsat15  \
24              NaN        NaN        NaN       NaN         NaN        NaN   
25              NaN        NaN        NaN       NaN         NaN        NaN   
26              NaN        NaN        NaN       NaN         NaN        NaN   
27              NaN        NaN        NaN       NaN         NaN        NaN   
28              NaN        NaN        NaN       NaN         NaN        NaN   
...             ...        ...        ...       ...         ...        ...   
133409      -1338.0     -597.0      -23.0       0.0         0.0        1.0   
133410      -1831.0     -816.0       -1.0       0.0         0.0        1.0   
133411      -1613.0     -209.0      -10.0       0.0         0.0        2.0   
133412      -1663.0     -845.0        1.0       0.0         0.0        1.0   
133413       -433.0     -182.0        4.0       0.0         0.0        3.0   

       opp_killsat15 opp_assistsat15 opp_deathsat15  
24               NaN             NaN            NaN  
25               NaN             NaN            NaN  
26               NaN             NaN            NaN  
27               NaN             NaN            NaN  
28               NaN             NaN            NaN  
...              ...             ...            ...  
133409           1.0             1.0            0.0  
133410           2.0             6.0            0.0  
133411           1.0             4.0            0.0  
133412           3.0             4.0            0.0  
133413           1.0             4.0            0.0  

[32520 rows x 130 columns]

In [74]:
# verification
tier1_player['playername'].head(15)

24       shanji
25          Aki
26        Creme
27         Able
28         COLD
29          New
30        Moyan
31       Feikun
32         Puff
33    Southwind
48       shanji
49          Aki
50        Creme
51         Able
52         COLD
Name: playername, dtype: object

#### Team Data

In [75]:
tier1_team = tier1.iloc[[i for i in range(len(tier1)) if i%12 in {10, 11}]]
tier1_team

gameid datacompleteness  \
34           8401-8401_game_1          partial   
35           8401-8401_game_1          partial   
58           8401-8401_game_2          partial   
59           8401-8401_game_2          partial   
82           8402-8402_game_1          partial   
...                       ...              ...   
133391  ESPORTSTMNT02_2965321         complete   
133402  ESPORTSTMNT03_2961731         complete   
133403  ESPORTSTMNT03_2961731         complete   
133414  ESPORTSTMNT03_2961733         complete   
133415  ESPORTSTMNT03_2961733         complete   

                                                url league  year   split  \
34      https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
35      https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
58      https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
59      https://lpl.qq.com/es/stats.shtml?bmid=8401    LPL  2022  Spring   
82      https://lpl.qq.com/es/stats.shtml?bmid=8402    LPL  2022  Spring   
...                                             ...    ...   ...     ...   
133391                                          NaN    LCS  2022  Summer   
133402                                          NaN    LLA  2023     NaN   
133403                                          NaN    LLA  2023     NaN   
133414                                          NaN    LLA  2023     NaN   
133415                                          NaN    LLA  2023     NaN   

        playoffs                 date  game  patch  ...  opp_csat15  \
34             0  2022-01-10 09:24:26     1  12.01  ...         NaN   
35             0  2022-01-10 09:24:26     1  12.01  ...         NaN   
58             0  2022-01-10 10:09:22     2  12.01  ...         NaN   
59             0  2022-01-10 10:09:22     2  12.01  ...         NaN   
82             0  2022-01-10 11:26:11     1  12.01  ...         NaN   
...          ...                  ...   ...    ...  ...         ...   
133391         1  2022-09-11 21:34:49     3  12.15  ...       466.0   
133402         0  2022-09-11 21:37:14     3  12.16  ...       546.0   
133403         0  2022-09-11 21:37:14     3  12.16  ...       506.0   
133414         0  2022-09-11 23:09:07     4  12.16  ...       485.0   
133415         0  2022-09-11 23:09:07     4  12.16  ...       514.0   

       golddiffat15 xpdiffat15 csdiffat15 killsat15 assistsat15 deathsat15  \
34              NaN        NaN        NaN       NaN         NaN        NaN   
35              NaN        NaN        NaN       NaN         NaN        NaN   
58              NaN        NaN        NaN       NaN         NaN        NaN   
59              NaN        NaN        NaN       NaN         NaN        NaN   
82              NaN        NaN        NaN       NaN         NaN        NaN   
...             ...        ...        ...       ...         ...        ...   
133391       -195.0     1700.0       21.0       3.0         5.0        4.0   
133402      -2304.0    -1702.0      -40.0       0.0         0.0        3.0   
133403       2304.0     1702.0       40.0       3.0         6.0        0.0   
133414       6878.0     2649.0       29.0       8.0        19.0        0.0   
133415      -6878.0    -2649.0      -29.0       0.0         0.0        8.0   

       opp_killsat15 opp_assistsat15 opp_deathsat15  
34               NaN             NaN            NaN  
35               NaN             NaN            NaN  
58               NaN             NaN            NaN  
59               NaN             NaN            NaN  
82               NaN             NaN            NaN  
...              ...             ...            ...  
133391           4.0             7.0            3.0  
133402           3.0             6.0            0.0  
133403           0.0             0.0            3.0  
133414           0.0             0.0            8.0  
133415           8.0            19.0            0.0  

[6504 rows x 130 columns]

In [76]:
# verification
tier1_team.loc[:, 'pick1':'pick5']

pick1      pick2     pick3      pick4         pick5
34          Jinx  Jarvan IV  Nautilus     Syndra          Gwen
35      Xin Zhao     Thresh  Aphelios        Vex           Jax
58          Jinx   Xin Zhao     Rakan     Rumble         Corki
59       Lee Sin      Leona     Ziggs  Gangplank  Twisted Fate
82          Jinx      Viego    Thresh      Corki        Graves
...          ...        ...       ...        ...           ...
133391    Wukong       Lulu     Sivir     Zilean          Gnar
133402     Sivir         Vi     Yuumi       Ornn        Viktor
133403    Wukong       Zeri      Ahri      Rakan        Aatrox
133414   Kalista     Wukong  Nautilus        Vex      Renekton
133415   Sejuani      Sylas        Vi   Aphelios    Tahm Kench

[6504 rows x 5 columns]

### Univariate Analysis

#### Number of Games Played in Each Patch
We can see that there were no games played in tier 1 leagues in 12.07, and there were only few games in 12.06, 12.08, and 12.16. It is because that most league spring playoffs happened in 12.05 and 12.06, so there was no game played in 12.07. The Mid-Seasonal Invitational(MSI) was hold during 12.08, so there was few games in 12.08 in tier 1 leagues, as we do not include the international competitions. Same reason apply to 12.16, most league summer playoffs happened in 12.15 and 12.16, and then the Worlds happened in 12.18.

In [78]:
fig1 = tier1_player.groupby('gameid').first()['patch'].value_counts().sort_index().plot(kind='bar')
fig1.update_layout(title='Number of Games Played in Each Patch',
                   xaxis_title='Patch',
                   yaxis_title='Number of games',
                   showlegend=False)
fig1.show()

#### Distribution of Game Length
For the following plot, we can see that the distribution of game length is skewed to the right, with the **median** of **31.39 minutes**.

In [79]:
# Median game length (minutes)
median_time = tier1_player.groupby('gameid').first()['gamelength'].apply(lambda x: x / 60).median()
median_time

31.391666666666666

In [80]:
fig2 = tier1_player.groupby('gameid').first()['gamelength'].apply(lambda x: x / 60).plot(kind='hist')
fig2.add_vline(x=median_time, line_dash="dash", line_color="red")
fig2.add_annotation(x=median_time, y=1.05, text="Median Game Length", showarrow=False, yref="paper")
fig2.update_layout(title='Distribution of Game Length',
                   xaxis_title='Game length (minutes)',
                   yaxis_title='Number of games',
                   showlegend=False)
fig2.show()

### Bivariate Analysis

#### Average Game Length in Each Tier 1 League
From the following boxplot and barplot, we can see that there is no significant game length between each league. One interesting fact is that **LCK**, known for best game strategy in the game with less teamfight, has the longest average game length, while **VCS**, known for its bloody and frequent teamfights, has the shortest average game length.

In [81]:
games = tier1_player.groupby('gameid').first()
games['gamelength'] = games['gamelength'] / 60
fig3 = px.box(data_frame=games, x='league', y='gamelength')
fig3.update_layout(title='Boxplot of Game Length in Each Tier 1 League',
                   xaxis_title='League',
                   yaxis_title='Game Length (minutes)')
fig3.show()

In [82]:
fig4 = games.groupby('league')['gamelength'].mean().sort_values(ascending=False).plot(kind='bar')
fig4.update_layout(title='Average Game Length in Each Tier 1 League',
                   xaxis_title='League',
                   yaxis_title='Average Game Length (minutes)',
                   yaxis_range=[0, 40], 
                   showlegend=False)
fig4.show()

#### Blue / Red Team Win Rate in Each Tier 1 League
From the table and grouped barplot below, we can see that in most leagues (except LCS, LLA, and VCS), **the win rate of blue side is higher than the win rate of red side.** Especially in PCS, the difference of the win rates between blue and red sides is about 0.20, which is a significant amount that can affect the result of the game.

In [83]:
win_rate = tier1_team.groupby(['league', 'side']).mean()['result'].unstack()
win_rate

side        Blue       Red
league                    
CBLOL   0.539095  0.460905
LCK     0.507495  0.492505
LCO     0.514151  0.485849
LCS     0.493464  0.506536
LEC     0.534979  0.465021
LJL     0.514019  0.485981
LLA     0.491979  0.508021
LPL     0.549618  0.450382
PCS     0.605166  0.394834
VCS     0.495356  0.504644

In [84]:
fig5 = px.bar(win_rate, x=win_rate.index, y=['Blue', 'Red'], barmode='group')
fig5.update_layout(title='Blue / Red Team Win Rate in Each Tier 1 League',
                   xaxis_title='League',
                   yaxis_title='Win Rate',
                   legend_title='Side')
fig5.show()

### Interesting Aggregates

#### Number of Games Played in Each Tier 1 League in 2022
From the plot below, we can see that LPL has the most games played in 2022. This is because they had 17 teams with Single Round Robin and BO3(Best of three) in regular seasons. [(Source)](https://lol.fandom.com/wiki/LPL/2022_Season) The second largest is LCK, which had 10 teams with Double Round Robin and BO3 in regular seasons. [(Source)](https://lol.fandom.com/wiki/LCK/2022_Season)

In [85]:
fig6 = tier1_player.groupby('gameid').first()['league'].value_counts().plot(kind='bar')
fig6.update_layout(title='Number of Games Played in Each Tier 1 League in 2022',
                  xaxis_title='League',
                  yaxis_title='Number of games',
                  showlegend=False)
fig6.show()

#### Most Picked Champions in Each Position
From the pivot table and the plot below, we can see the most played 10 picked champions in each position.

In [86]:
champ_picked = tier1_player.pivot_table(index='position', columns='champion', aggfunc='size', fill_value=0).T
champ_picked = (champ_picked.reindex(columns=['top', 'jng', 'mid', 'bot', 'sup'])
                .rename(columns={'top': 'Top', 'jng': 'Jungle', 'mid': 'Mid', 'bot': 'Bot', 'sup': 'Support'})
                )
champ_picked

position  Top  Jungle  Mid  Bot  Support
champion                                
Aatrox    224       0    1    0        0
Ahri        0       0  886    0        0
Akali     144       0  203    0        0
Akshan     21       0    2    0        0
Alistar     0       0    0    0      270
...       ...     ...  ...  ...      ...
Zeri        2       0    2  818        0
Ziggs       0       0    5   38        0
Zilean      5       0   40    0       35
Zoe         0       0  147    0        0
Zyra        0       0    0    0        2

[152 rows x 5 columns]

In [87]:
for col in champ_picked.columns:
    fig = champ_picked[col].sort_values(ascending=False).iloc[:10].plot(kind='bar', title=col)
    fig.update_layout(title=f'Most Picked 10 Champions in {col}',xaxis_title='Champion', yaxis_title='Number of picks', showlegend=False)
    fig.show()

## Step 3: Assessment of Missingness

We have two datasets: `tier1_player` and `tier1_team`. We will only use `tier1_team` for later parts, and we will only keep columns that we are interested.

In [106]:
data = tier1_team[['gameid', 'league', 'year', 'split', 'playoffs', 'date', 'game', 'patch', 'side', 'teamname', 'teamid',
            'ban1', 'ban2', 'ban3', 'ban4', 'ban5', 'pick1', 'pick2', 'pick3', 'pick4', 'pick5', 
            'gamelength', 'result', 'kills', 'deaths', 'assists']]
data.head()

gameid league  year   split  playoffs                 date  \
34  8401-8401_game_1    LPL  2022  Spring         0  2022-01-10 09:24:26   
35  8401-8401_game_1    LPL  2022  Spring         0  2022-01-10 09:24:26   
58  8401-8401_game_2    LPL  2022  Spring         0  2022-01-10 10:09:22   
59  8401-8401_game_2    LPL  2022  Spring         0  2022-01-10 10:09:22   
82  8402-8402_game_1    LPL  2022  Spring         0  2022-01-10 11:26:11   

    game  patch  side            teamname  ...     pick1      pick2     pick3  \
34     1  12.01  Blue           Oh My God  ...      Jinx  Jarvan IV  Nautilus   
35     1  12.01   Red  ThunderTalk Gaming  ...  Xin Zhao     Thresh  Aphelios   
58     2  12.01  Blue           Oh My God  ...      Jinx   Xin Zhao     Rakan   
59     2  12.01   Red  ThunderTalk Gaming  ...   Lee Sin      Leona     Ziggs   
82     1  12.01  Blue     FunPlus Phoenix  ...      Jinx      Viego    Thresh   

        pick4         pick5 gamelength result kills deaths assists  
34     Syndra          Gwen       1365      1    13      6      35  
35        Vex           Jax       1365      0     6     13      11  
58     Rumble         Corki       1444      1    22      9      40  
59  Gangplank  Twisted Fate       1444      0     8     22      16  
82      Corki        Graves       1893      1    12      8      25  

[5 rows x 26 columns]

From the following code block, we can see that missingness values exist in `split`, `ban`, `pick` columns, so we are going to examine further.

In [107]:
data.isnull().sum()

gameid          0
league          0
year            0
split         166
playoffs        0
date            0
game            0
patch           0
side            0
teamname        0
teamid          0
ban1            7
ban2            4
ban3           14
ban4           16
ban5           28
pick1          42
pick2          42
pick3          42
pick4          42
pick5          42
gamelength      0
result          0
kills           0
deaths          0
assists         0
dtype: int64

Since the pick1 - pick5 columns have the same number of missing values, it looks that they are missing together.

From the dataframe below, there is only one game (two rows, blue and red) that has nan in pick columns is from LPL 2022 Spring Regular Season, but the others are from VCS 2022 Summer Regular Season.

In [119]:
data[data['pick1'].isnull()].groupby(['league', 'year', 'split', 'playoffs']).size()

league  year  split   playoffs
LPL     2022  Spring  0            2
VCS     2022  Summer  0           40
dtype: int64

From the below, we can see that those ban columns usually only have one nan value in the row (except for 8473-8473_game_1, which is the game from LPL mentioned above), so we would like to argue that those nan values is **MCAR**, because it is allowed to have empty ban, which become nan values in our data. We would like to carry out permutation tests to test if the `ban1` to `ban5` columns are dependent to the `league` column.

In [118]:
data[data.loc[:,'ban1': 'ban5'].isnull().any(axis=1)].loc[:, :'ban5']

gameid league  year    split  playoffs  \
1487    ESPORTSTMNT01_2690705    LCK  2022   Spring         0   
2507    ESPORTSTMNT01_2691557    LCK  2022   Spring         0   
10642   ESPORTSTMNT02_2552213  CBLOL  2022  Split 1         0   
14854   ESPORTSTMNT02_2576641    LCO  2022  Split 1         0   
19786   ESPORTSTMNT02_2558786    VCS  2022   Spring         0   
24539        8468-8468_game_1    LPL  2022   Spring         0   
25738        8473-8473_game_1    LPL  2022   Spring         0   
25739        8473-8473_game_1    LPL  2022   Spring         0   
26242   ESPORTSTMNT03_2547408    PCS  2022   Spring         0   
26327   ESPORTSTMNT03_2547435    LCK  2022   Spring         0   
28811   ESPORTSTMNT03_2549195    LCO  2022  Split 1         0   
29050        8746-8746_game_2    LPL  2022   Spring         0   
31186   ESPORTSTMNT01_2731398    LCS  2022   Spring         0   
32447   ESPORTSTMNT03_2590257    PCS  2022   Spring         0   
33131        8759-8759_game_2    LPL  2022   Spring         0   
34858   ESPORTSTMNT01_2734124    LCK  2022   Spring         0   
37271        8767-8767_game_1    LPL  2022   Spring         0   
43510   ESPORTSTMNT01_2810127    LJL  2022   Spring         0   
49499   ESPORTSTMNT03_2631825  CBLOL  2022  Split 1         0   
49870   ESPORTSTMNT02_2663198    LCO  2022  Split 1         0   
52822        8795-8795_game_2    LPL  2022   Spring         0   
53723   ESPORTSTMNT03_2633616  CBLOL  2022  Split 1         1   
54359   ESPORTSTMNT01_2840075    VCS  2022   Spring         0   
54443   ESPORTSTMNT03_2653794  CBLOL  2022  Split 1         1   
60946   ESPORTSTMNT03_2657112  CBLOL  2022  Split 1         1   
65519   ESPORTSTMNT03_2711052    VCS  2022   Summer         0   
73186   ESPORTSTMNT02_2785385    LCO  2022  Split 2         0   
76090   ESPORTSTMNT02_2788476    LLA  2022  Closing         0   
76523        8952-8952_game_2    LPL  2022   Summer         0   
79295        8958-8958_game_3    LPL  2022   Summer         0   
80231   ESPORTSTMNT03_2780993    LEC  2022   Summer         0   
86363   ESPORTSTMNT03_2796815    LLA  2022  Closing         0   
89566        8986-8986_game_2    LPL  2022   Summer         0   
94691   ESPORTSTMNT01_2982943    LCO  2022  Split 2         0   
98135   ESPORTSTMNT02_2911303    LCO  2022  Split 2         0   
98195        9280-9280_game_3    LPL  2022   Summer         0   
99035   ESPORTSTMNT02_2911639    LCO  2022  Split 2         0   
107735  ESPORTSTMNT05_2980556    LCK  2022   Summer         0   
109451  ESPORTSTMNT03_2892347    LCO  2022  Split 2         0   
110902  ESPORTSTMNT06_2531032    LCK  2022   Summer         0   
111634       9315-9315_game_1    LPL  2022   Summer         0   
112715  ESPORTSTMNT03_2884198    PCS  2022   Summer         0   
115162  ESPORTSTMNT03_2885491    LCO  2022  Split 2         0   
115211  ESPORTSTMNT03_2885497    LCO  2022  Split 2         0   
117983       9331-9331_game_2    LPL  2022   Summer         0   
119723  ESPORTSTMNT03_2920601    VCS  2022   Summer         0   
120179  ESPORTSTMNT01_3092256    LCS  2022   Summer         0   
120370  ESPORTSTMNT01_3092439    LCO  2022  Split 2         0   
120791  ESPORTSTMNT01_3092718    LCO  2022  Split 2         0   
124030  ESPORTSTMNT02_2960376    LCO  2022  Split 2         1   
124054  ESPORTSTMNT02_2960378    LCO  2022  Split 2         1   

                       date  game  patch  side                       teamname  \
1487    2022-01-12 10:07:10     1  12.01   Red               Kwangdong Freecs   
2507    2022-01-14 13:00:25     3  12.01   Red                             T1   
10642   2022-01-29 18:14:54     1  12.02  Blue                KaBuM! e-Sports   
14854   2022-02-07 06:43:48     1  12.02  Blue                       Gravitas   
19786   2022-02-13 10:43:35     2  12.02  Blue                   SBTC Esports   
24539   2022-02-18 07:23:05     1  12.02   Red                      Rare Atom   
25738   2022-02-19 11:34:46     1  12.02  Blue                      JD Gami

### Permutation Tests
- test statistic: the proportion of missing value in each league

In [138]:
ban_data = data.loc[:, ['league','ban1', 'ban2', 'ban3', 'ban4', 'ban5']]
for col in ban_data.columns[1:]:
    ban_data[col] = ban_data[col].isna()
ban_data

league   ban1   ban2   ban3   ban4   ban5
34        LPL  False  False  False  False  False
35        LPL  False  False  False  False  False
58        LPL  False  False  False  False  False
59        LPL  False  False  False  False  False
82        LPL  False  False  False  False  False
...       ...    ...    ...    ...    ...    ...
133391    LCS  False  False  False  False  False
133402    LLA  False  False  False  False  False
133403    LLA  False  False  False  False  False
133414    LLA  False  False  False  False  False
133415    LLA  False  False  False  False  False

[6504 rows x 6 columns]

In [188]:
pt =ban_data[['league', 'ban1']].pivot_table(index='league', columns='ban1', aggfunc='size', fill_value=0)
pt[True] / pt.sum(axis=1)

league
CBLOL    0.000000
LCK      0.000000
LCO      0.004717
LCS      0.000000
LEC      0.000000
LJL      0.000000
LLA      0.000000
LPL      0.001272
PCS      0.000000
VCS      0.004644
dtype: float64

In [171]:
ban_data[['league', 'ban1']].pivot_table(index='league', columns='ban1', aggfunc='size', fill_value=0)

ban1    False  True
league             
CBLOL     486     0
LCK       934     0
LCO       422     2
LCS       612     0
LEC       486     0
LJL       428     0
LLA       374     0
LPL      1570     2
PCS       542     0
VCS       643     3

In [178]:
shuffled = ban_data.copy()[['league', 'ban1']]
shuffled['ban1'] = np.random.permutation(shuffled['ban1'])
pivoted = shuffled.pivot_table(index='ban1', columns='league', aggfunc='size', fill_value=0)
pivoted = pivoted / pivoted.sum()
# tvd = pivoted.diff(axis=1).iloc[:, -1].abs().sum() / 2

In [179]:
pivoted

league  CBLOL       LCK  LCO       LCS       LEC       LJL  LLA       LPL  \
ban1                                                                        
False     1.0  0.998929  1.0  0.998366  0.997942  0.995327  1.0  0.998728   
True      0.0  0.001071  0.0  0.001634  0.002058  0.004673  0.0  0.001272   

league  PCS  VCS  
ban1              
False   1.0  1.0  
True    0.0  0.0

In [ ]:
n_repetitions = 500
shuffled = ban_data.copy()

tvds = []
for _ in range(n_repetitions):
    shuffled['ban1'] = np.random.permutation(shuffled['ban1'])
    
    # Computing and storing the TVD.
    pivoted = (
        shuffled
        .pivot_table(index='gender', columns='child_missing', aggfunc='size')
    )
    
    pivoted = pivoted / pivoted.sum()
    
    tvd = pivoted.diff(axis=1).iloc[:, -1].abs().sum() / 2
    tvds.append(tvd)

## Step 4: Hypothesis Testing

In [ ]:
# TODO

## Step 5: Framing a Prediction Problem

In [ ]:
# TODO

## Step 6: Baseline Model

In [ ]:
# TODO

## Step 7: Final Model

In [ ]:
# TODO

## Step 8: Fairness Analysis

In [ ]:
# TODO